src/main.cpp

```diff
@@ -497,10 +497,7 @@ int llama_generate(struct llama_context_wrapper * ctx_w, gpt_params params, py::
                 std::string line;
                 bool another_line = true;
                 do {
-                    if (!std::getline(std::cin, line)) {
-                        // input stream is bad or EOF received
-                        return 0;
-                    }
+                    line = grab_text_callback().cast<std::string>();
```

Comment:
Is there a way to let the callback signal `EOF` then check it here so that the function can terminate gracefully?

Reply:
@Kasimir123 Yeah You are right.
I had a similar idea but was working with a different approach. I want to have more control over the inference so I decided to re-implement the logic in a similar way. 
What @nuance1979 is asking for (and many other people) is difficult to do from the callback. 

@Kasimir123 let me know if you have any idea on how to signal `EOF`, otherwise I will merge the PR ?

Reply:
I can think of two ways to signal EOF:

1. Use a special string (e.g., `"[EOF]"`) or character (e.g., `"\x03"`);
2. Let the callback return a `Tuple[string, bool]` where the boolean means EOF.

How do you think?

Reply:
Of course, we could merge this PR first for now and continue this discussion in another PR. @Kasimir123 @abdeladim-s

Reply:
@Kasimir123 is not replying, so I don't know if he is still working on it. 

@nuance1979, I already implemented a different logic for interactive mode. So, this approach won't be necessary any more.

Reply:
@nuance1979 @abdeladim-s I managed to add EOF file handling by adding a check to see if the callback result was of type string. This means that if you want end of file you can return None and it will correctly exit the program. To test this change change your callback to:

```py
def grab_text_callback() -> str:
    inpt = input()
    if inpt == "END":
        return None
    return inpt + "\n"
```